pkg/common/peertracker/listener_posix.go

```diff
@@ -34,8 +43,24 @@ func (lf *ListenerFactory) listenUnix(network string, laddr *net.UnixAddr) (*Lis
 	}
 
 	return &Listener{
-		l:       l,
+		l:       limitedListener,
 		Tracker: tracker,
 		log:     lf.Log,
 	}, nil
 }
+
+func (lf *ListenerFactory) limitListenerBelowRlimit(l net.Listener) (net.Listener, error) {
+	var (
+		concurrency   int
+		rlimit        syscall.Rlimit
+		rlimitPercent uint64 = 99
+	)
+	if err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rlimit); err != nil {
+		return l, err
+	}
+	concurrency = int(rlimit.Cur * rlimitPercent / 100)
```

Comment:
As written, this is essentially proof of concept that we could potentially ratelimit/loadshed in agent, but not representative of an optimum algorithm to calculate the actual concurrency threshold.